3rdParty/fuerte/src/HttpConnection.cpp

```diff
@@ -562,6 +623,18 @@ void HttpConnection<ST>::drainQueue(const fuerte::Error ec) {
   }
 }
 
+// Lease this connection (cancel idle alarm)
+template <SocketType ST>
+bool HttpConnection<ST>::lease() {
+  int exp = 0;
+  if (!this->_leased.compare_exchange_strong(exp, 1) ||
+      this->_state == Connection::State::Failed) {
```

Comment:
In principle, yes, since another thread could call `cancel` on the connection object. But then it is OK, that in the TLS case the next request fails. The iothread itself can only set the state to `Failed` if a real error occurs (in which case it is OK to run into an error), or if the idle timeout alarm goes off. This code guards against the idle alarm going off despite the fact that the connection was already leased, since we **first** compare exchange from 0 to 1 and then check again that the connection is not Failed. Both accesses happen with `std::memory_order_seq_cst` and the check for `_leased` in the idle alarm handler happens before `shutdownConnection` is called. Therefore, I think this is OK here.

Reply:
More comments added.